x/stakeibc/keeper/msg_server_instant_redeem_stake.go

```diff
@@ -0,0 +1,112 @@
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+	"github.com/Stride-Labs/stride/v5/x/stakeibc/types"
+)
+
+func (k msgServer) InstantRedeemStake(goCtx context.Context, msg *types.MsgInstantRedeemStake) (*types.MsgInstantRedeemStakeResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	k.Logger(ctx).Info(fmt.Sprintf("fast unbond: %s", msg.String()))
+
+	// get our addresses, make sure they're valid
+	sender, err := sdk.AccAddressFromBech32(msg.Creator)
+	if err != nil {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "creator address is invalid: %s. err: %s", msg.Creator, err.Error())
+	}
+
+	// then make sure host zone is valid
+	hostZone, found := k.GetHostZone(ctx, msg.HostZone)
+	if !found {
+		return nil, sdkerrors.Wrapf(types.ErrInvalidHostZone, "host zone is invalid: %s", msg.HostZone)
+	}
+
+	// safety check: redemption rate must be within safety bounds
+	rateIsSafe, err := k.IsRedemptionRateWithinSafetyBounds(ctx, hostZone)
+	if !rateIsSafe || (err != nil) {
+		errMsg := fmt.Sprintf("IsRedemptionRateWithinSafetyBounds check failed. hostZone: %s, err: %s", hostZone.String(), err.Error())
+		return nil, sdkerrors.Wrapf(types.ErrRedemptionRateOutsideSafetyBounds, errMsg)
+	}
+
+	// construct desired unstaking amount from host zone
+	nativeAmount := sdk.NewDecFromInt(msg.Amount).Mul(hostZone.RedemptionRate).RoundInt()
+	// Redemption amount must not be greater than staked balance in that zone.
+	if nativeAmount.GT(hostZone.StakedBal) {
+		return nil, sdkerrors.Wrapf(types.ErrInvalidAmount, "cannot unstake an amount g.t. staked balance on host zone: %v", msg.Amount)
+	}
+	// safety checks on the coin
+	// 	- Redemption amount must be positive
+	if !nativeAmount.IsPositive() {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidCoins, "amount must be greater than 0. found: %v", msg.Amount)
+	}
+
+	// get the native Ibc coin to return to user
+	ibcDenom := hostZone.GetIbcDenom()
+	nativeCoin := sdk.NewCoin(ibcDenom, nativeAmount)
+
+	stDenom := types.StAssetDenomFromHostZoneDenom(hostZone.HostDenom)
+	// 	- Creator owns at least "amount" stAssets
+	balance := k.bankKeeper.GetBalance(ctx, sender, stDenom)
+	k.Logger(ctx).Info(fmt.Sprintf("Redemption issuer IBCDenom balance: %v%s", balance.Amount, balance.Denom))
+	if balance.Amount.LT(msg.Amount) {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidCoins, "balance is lower than redemption amount. redemption amount: %v, balance %v: ", msg.Amount, balance.Amount)
+	}
+
+	// Find and subtract this amount from a deposit record if it is big enough
+	depositRecords := k.RecordsKeeper.GetAllDepositRecord(ctx)
+	pendingDepositRecords := utils.FilterDepositRecords(depositRecords, func(record recordstypes.DepositRecord) (condition bool) {
+		return record.Status == recordstypes.DepositRecord_TRANSFER_QUEUE && record.HostZoneId == hostZone.ChainId
+	})
+	totalPendingDeposits := utils.SumDepositRecords(pendingDepositRecords)
```

Comment:
I believe we should only ever have 1 DepositRecord with status TRANSFER_QUEUE per HostZone at a time, so might not need to sum here, and might be able to replace the loop on ln 75 with a simple subtraction. @asalzmann can you check me on this?

Reply:
I did it that way initially, and changed it to this upon @asalzmann's request and comment above.  I reviewed the code, and it does look like there can be multiple entries if they timeout.